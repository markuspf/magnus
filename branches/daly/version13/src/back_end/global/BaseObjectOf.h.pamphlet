\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/BaseObjectOf.h}
\author{Stephane Collart}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{include}
<<include>>=
#include "PureRep.h"
#include "ObjectOf.h"

@
\section{template<class Rep> class BaseObjectOf}
<<template<class Rep> class BaseObjectOf>>=
template<class Rep> class BaseObjectOf : private ObjectOf<Rep>
{
// This assumes that Rep is derived from PureRep
public:
  // copy constructor provided by compiler
  // destructor provided by compiler
  // assignment operator provided by compiler

protected:
  // Inherited from base; adjust access:
  ObjectOf<Rep>::look;
  ObjectOf<Rep>::change;
  ObjectOf<Rep>::enhance;
  ObjectOf<Rep>::acquireRep;
  BaseObjectOf( Rep* newrep ) : ObjectOf<Rep>( newrep) { }

private:
	// Dummy Forcing Member:
	void force_derivation( ) { const PureRep* rc = look(); }
	// With this member PureRep is forced to be 
        // an accessible base class of Rep
};

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/BaseObjectOf.h
#ifndef _BASE_OBJECT_OF_H_
#define _BASE_OBJECT_OF_H_

<<license>>
<<include>>
<<template<class Rep> class BaseObjectOf>>
#endif

@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
