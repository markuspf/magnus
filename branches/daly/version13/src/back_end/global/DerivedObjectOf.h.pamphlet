\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/DerivedObjectOf.h}
\author{Stephane Collart}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{class DerivedObjectOf>
<<class DerivedObjectOf>>=
template <class Base, class Rep> class DerivedObjectOf : public Base
{

// This class assumes Base is itself an ObjectOf kind of class, i.e.
// 1) Base has members look(), enhance(), change() returning some BaseRep* type
// 2) Base has a constructor Base( BaseRep* );
// 3) BaseRep is directly or indirectly derived from PureRep
// 4) Rep is derived from BaseRep

public:
	// copy constructor provided by compiler
	// destructor provided by compiler
	// assignment operator provided by compiler

protected:
	// Shadowed representation accessors:
	const Rep* look( ) const { return (Rep*)Base::look(); }
	Rep* enhance( ) const { return (Rep*)Base::enhance(); }
	Rep* change( ) { return (Rep*)Base::change(); }

	DerivedObjectOf( Rep* newrep ) : Base( newrep ) { }
	// for base class initialisation

	DerivedObjectOf( const Base& b ) : Base( b ) { }
	// for base class intialisation when cast conversion
        // to derived is desired 

};

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/DerivedObjectOf.h
#ifndef _DERIVED_OBJECT_OF_H_
#define _DERIVED_OBJECT_OF_H_

<<license>>
<<class DerivedObjectOf>>
#endif
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
