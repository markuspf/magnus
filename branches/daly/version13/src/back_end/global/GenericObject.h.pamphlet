\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/GenericObject.h}
\author{Stephane Collart}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{include}
<<include>>=
#include "PureRep.h"
#include "BaseObjectOf.h"
#include "Type.h"

@
\section{struct GenericRep}
a common root class for all object-level (see PureObject.h) classes
with a coherent type scheme.

See PureObject

In addition, this class provides explicit type query handles; classes
derived from this class will normally derive *publicly*, to permit
assignment by non-friend
<<struct GenericRep>>=
struct GenericRep : public PureRep
{
	// default constructor provided by compiler
	// copy constructor provided by compiler
	// assignment operator disabled

	static const Type theGenericObjectType;
	// every representation class derived from this 
        // should carry a similarly defined type field

	static Type type() { return theGenericObjectType; }
	// every representation class derived from this 
        // should carry a similarly defined type member

	virtual Type actualType() const { return type(); }
	// this member serves as the root of all type querying members in this
	// type scheme
	// every representation class derived from this should override this
	// member similarly to return the actual type of the class

	PureRep* clone( ) const { return new GenericRep(*this); }

};

@
\section{class GenericObject}
<<class GenericObject>>=
class GenericObject : protected BaseObjectOf<GenericRep>
{
public:
	GenericObject( ) : BaseObjectOf<GenericRep>( new GenericRep ) { }

	// copy constructor provided by compiler
	// destructor provided by compiler
	// both of the preceding ref-counting compatible
	// assignment operator provided by compiler
	// ref-counting compatible

	static Type type() { return GenericRep::type(); }
	// this `type querying' member is provided solely as an example, 
        // since this class is pseudo-abstract (ie. its representation 
        // class is abstract) and no actual objects of this type will 
        // ever be created

	Type actualType() const { return look()->actualType(); }

protected:
	// look(), change() and enhance() inherited

	// Special wrapping constructor to wrap new representations (returned
	// by eg. delegated methods) and for base initialisation by derived
	// classes:

	GenericObject( GenericRep* newrep ) : BaseObjectOf<GenericRep>(newrep) { }

};

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/GenericObject.h
#ifndef _GENERIC_OBJECT_H_
#define _GENERIC_OBJECT_H_


<<include>>
<<license>>
<<struct GenericRep>>
<<class GenericObject>>
#endif
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
