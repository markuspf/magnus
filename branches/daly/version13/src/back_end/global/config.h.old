// Copyright (C) 1994 The New York Group Theory Cooperative
// See magnus/doc/COPYRIGHT for the full notice.

// Contents: Configuration variables set up by the config script
//
// Principal Author: Jon Lennox
//
// Status: Useable.
//
// Revision History:
//

#ifndef _CONFIG_H_
#define _CONFIG_H_

#define MAGNUS_HOME "/usr/bormotov/magnus"
// We keep this compiled-in value to ease debugging.
// Use MagnusHome::magnusHome() in back_end/general/ for
// the Magnus root directory.

#define LINUX

#define MAGNUS_TMP "/tmp"

#endif /* _CONFIG_H_ */
