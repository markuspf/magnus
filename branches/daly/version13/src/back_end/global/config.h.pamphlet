\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/config.h}
\author{Jon Lennox}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/config.h
#ifndef _CONFIG_H_
#define _CONFIG_H_

<<license>>

#define MAGNUS_HOME "/usr/local/magnus"
// We keep this compiled-in value to ease debugging.
// Use MagnusHome::magnusHome() in back_end/general/ for
// the Magnus root directory.

#define LINUX

#define MAGNUS_TMP "/tmp"

#endif /* _CONFIG_H_ */
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
