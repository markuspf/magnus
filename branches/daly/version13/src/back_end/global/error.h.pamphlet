\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/error.h}
\author{Roger Needham}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{error}
<<error>>=
#ifdef __GNUG__
//@jml This tells g++ that error does not return.
inline volatile void error(const char *msg) {
#else
inline void error(const char* msg) {
#endif

  cout << "fatal_error_ {" << msg << '}' << endl << flush;
  // Try to shove the message directly out to the front end
  cerr << endl << "Fatal error: " << msg << endl;
  while ( 1 );
  // So we can get a stack backtrace when debugging
}

@
\section{warn}
<<warn>>=
inline void warn(const char* msg) {
  cerr << endl << "Warning: " << msg << endl;
}

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/error.h
#ifndef _ERROR_H_
#define _ERROR_H_

<<license>>
<<error>>
<<warn>>
#endif
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
