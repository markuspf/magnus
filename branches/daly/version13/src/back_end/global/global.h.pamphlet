\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/global.h}
\author{Roger Needham}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{include}
<<include>>=
#include <stream.h>
#include <iostream.h>
#include <fstream.h>
#include <strstream.h>
//@rn #include <string.h>
#include <ctype.h>
#include <stdlib.h>
#include <math.h>

#include "error.h"
#include "IPC.h"
#include "Trichotomy.h"

@
\section{global.h}
Contents: Global declarations to be included by all other header files.
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/global.h
#ifndef _GLOBAL_H_
#define _GLOBAL_H_

<<include>>
<<license>>

//#define NULL 0

#define SAFETY 2

#define TRUE  1
#define FALSE 0

typedef int Bool;

#define INFINITE 0


//------ global function prototypes, variables, inline functions ------------//

inline int min(int i, int j) { return ((i<j) ? i : j); }
inline int max(int i, int j) { return ((i>j) ? i : j); }
inline int abs(int i) { return ((i < 0) ? -i : i); }
inline int sign(int i) { return (i == 0) ? 0 : ( (i > 0) ? 1 : -1 ); }
#endif
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
