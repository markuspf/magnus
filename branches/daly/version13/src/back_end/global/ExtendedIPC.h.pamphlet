\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/ExtendedIPC.h}
\author{Dmitry Bormotov}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{include}
<<include>>=
#include "global.h"
#include "Integer.h"
#include "Rational.h"

@
\section{Output Operators}
<<Output Operators>>=
inline ostream& operator < ( ostream& ostr, const Integer& n )
{
  ostr << n << ' ';
  return ostr;
}

inline ostream& operator < ( ostream& ostr, const Rational& r )
{
  ostr << r << ' ';
  return ostr;
}

@
\section{Input Operators}
<<Input Operators>>=
inline istream& operator > ( istream& istr, Integer& n )
{
  istr >> n;
  return istr;
}

inline istream& operator > ( istream& istr, Rational& r )
{
  istr >> r;
  return istr;
}

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/ExtendedIPC.h
#ifndef _EXTENDED_IPC_H_
#define _EXTENDED_IPC_H_

<<include>>
<<license>>
<<Output Operators>>
<<Input Operators>>
#endif

@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
