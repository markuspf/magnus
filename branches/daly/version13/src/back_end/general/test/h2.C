// Contents: The version of merge sort which runs insertion sort
//           when the number of elements to be sorted is <= k.
//           Time measurements are taken for different values of k.
//
// Principal Author: Dmitry Bormotov
//
// Status: Completed
//
// Revision History:
//

#include <stdio.h>
#include <signal.h>
#include <sys/time.h>
#include <unistd.h>
#include <fstream.h>
#include "global.h"
#include "RandomNumbers.h"

void swap( int* A, int i, int j );
int findLargest(int* A, int heapSize, int d, int i);
void printArray(int* A, int ASize);


// Definitions:
//
//   A[0 .. ASize-1] - array of integers, represents a heap
//   d - arity of the heap ( d >= 2 )
//   heapSize - size of the heap, always <= ASize 
//   Subtree(i) - subtree rooted at node i
//   Children(i) = { x: x belongs to the intersection of
//   {d*i+1,...,d*i+d} and {x < heapSize}  


void heapify( int* A, int heapSize, int d, int a )
// pre:  A[0 .. heapSize-1], 0 <= a < heapSize 
//       For all x, x is in Children(a) -> Subtree(x) is a heap
//
// post: Subtree(a) is a heap
{
  int i = a;
  int largest = findLargest(A,heapSize,d,i);

  // * Invariant: 
  // All nodes of Children(a) except i satisfy the heap property.
  // Node i may or may not satisfy the heap property.
  // largest is the node of the largest value among A[i] 
  // and values of children of i.

  while( largest != i ) {
    swap(A,i,largest);
    i = largest;
    largest = findLargest(A,heapSize,d,i);
  }
}

void swap( int* A, int i, int j )
// swaps values of A[i] and A[j] 
{
 int r = A[i]; 
 A[i] = A[j]; 
 A[j] = r;
}

int findLargest(int* A, int heapSize, int d, int i)
// finds the largest value among A[i] and values of children of i
// and returns its node in the heap
{
  int maxInd = i;
  for( int j = d*i+1; j <= d*i+d; ++j )
    if( j < heapSize && A[maxInd] < A[j] ) 
      maxInd = j;
  return maxInd;
}

void buildHeap( int* A, int ASize, int d, int& heapSize )
{
  heapSize = ASize;
  for( int i = ASize - 1; i >= 0; --i )
    heapify(A,heapSize,d,i);
}

void heapSort(int* A, int ASize, int d)
{
  int heapSize;
  buildHeap(A,ASize,d,heapSize);
  for( int i = ASize-1; i >= 1; --i ) {
    swap(A,0,i);
    --heapSize;
    heapify(A,heapSize,d,0);
  }
}

void printArray(int* A, int ASize)
{
  for( int i = 0; i < ASize; ++i )
    cout << " " << A[i];
  cout << endl;
}

double diffTime(timeval t2, timeval t1)
// computes the time diffirence between t2 and t1 
{
  double result;
  long s = t2.tv_sec - t1.tv_sec;
  result = (t2.tv_usec - t1.tv_usec) * 0.000001 + s;
  //if( t2.tv_usec < t1.tv_usec ) --result;
  return result;
}

int extractMax( int* A, int& heapSize, int d )
{
  if( heapSize < 1 )
    error("int extractMax( int* A, int heapSize, int d ) : "
	  "heap underflow");
  int max = A[0];
  A[0] = A[heapSize-1];
  --heapSize;
  heapify(A,heapSize,d,0);
  return max;
}

int main ( )
{
  timeval t1,t2;
  int maxSize = 1000000;
  int A[maxSize];
  const int ASize = 1000000;
  UniformRandom r;
  int heapSize;

  for( int d = 2; d <= 20; ++d ) {
    
    for( int i = 0; i < ASize; ++i )
      A[i] = ASize-i-1;
    //A[i] = int(r.rand() * ASize);

    buildHeap(A,ASize,d,heapSize);
    
    gettimeofday(&t1, NULL);

    for( int i = 0; i < 100000; ++i )
      extractMax(A,heapSize,d);

    gettimeofday(&t2, NULL);

    double dif = diffTime(t2,t1);
    cout << "d = " << d << "   time elapsed: " << dif << endl;
  }
}

