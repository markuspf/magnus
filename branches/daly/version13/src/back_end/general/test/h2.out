The running time for the extractMax: 
d - arity of the heap, 
time is in seconds, 
the array contained 1 million numbers in reversed order, 
namely 999999, 999998, ..., 0,
extractMax has been called 100000 times: 

d = 2    time elapsed: 0.563652
d = 3    time elapsed: 0.436172
d = 4    time elapsed: 0.393798    The best time
d = 5    time elapsed: 0.484299
d = 6    time elapsed: 0.461851
d = 7    time elapsed: 0.464351
d = 8    time elapsed: 0.479125
d = 9    time elapsed: 0.47891
d = 10   time elapsed: 0.469805
d = 11   time elapsed: 0.482382
d = 12   time elapsed: 0.493437
d = 13   time elapsed: 0.500535
d = 14   time elapsed: 0.50854
d = 15   time elapsed: 0.512837
d = 16   time elapsed: 0.513829
d = 17   time elapsed: 0.512724
d = 18   time elapsed: 0.51666
d = 19   time elapsed: 0.529707
d = 20   time elapsed: 0.544394


Same experiment except that the array is filled with random values
from 0 to 999999 (still d = 4 is the best):

d = 2    time elapsed: 0.601835
d = 3    time elapsed: 0.503277
d = 4    time elapsed: 0.453177    The best time
d = 5    time elapsed: 0.544464
d = 6    time elapsed: 0.52148
d = 7    time elapsed: 0.524222
d = 8    time elapsed: 0.515534
d = 9    time elapsed: 0.544836
d = 10   time elapsed: 0.544615
d = 11   time elapsed: 0.563901
d = 12   time elapsed: 0.572443
d = 13   time elapsed: 0.590613
d = 14   time elapsed: 0.600558
d = 15   time elapsed: 0.604601
d = 16   time elapsed: 0.592821
d = 17   time elapsed: 0.627564
d = 18   time elapsed: 0.646315
d = 19   time elapsed: 0.665627
d = 20   time elapsed: 0.680557
